### PR TITLE
Add support of power management for stm32 CRYP peripheral

### PR DESCRIPTION
**Pull request by tbourgoi**
This pull request adds the support of power management for CRYP peripheral.
We also remove the requirement on the reset binding.

### CODE REVIEW

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 1]
```diff
@@ -1266,10 +1313,7 @@ static TEE_Result stm32_cryp_probe(const void *fdt, int node,
 	if (clk_enable(cryp_pdata.clock))
 		panic();
 
-	if (rstctrl && rstctrl_assert_to(cryp_pdata.reset, TIMEOUT_US_1MS))
-		panic();
-
-	if (rstctrl && rstctrl_deassert_to(cryp_pdata.reset, TIMEOUT_US_1MS))
+	if (stm32_cryp_reset())
```
Comment (jforissier) [inline core/drivers/crypto/stm32/stm32_cryp.c:1316]: This should be part of a separate commit since it also introduces a delay, so it's not just code refactoring.

**Review by jforissier — COMMENTED** [review]
> Remove panic during probe when reset bindings is not found because it's

`...when "resets" property is not found...` perhaps? Plus one more comment below.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 2]
```diff
@@ -84,6 +85,7 @@
 
 #define CRYP_TIMEOUT_US			1000000U
 #define TIMEOUT_US_1MS			1000U
+#define CRYP_RESET_DELAY		U(2)
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:88]: prefer explicitly state that the value is a microsecond time ref, 
for example s/CRYPT_RESET_DELAY/CRYPT_RESET_DELAY_US/g

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 3]
```diff
@@ -1231,6 +1233,51 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
 	return res;
 }
 
+static TEE_Result stm32_cryp_reset(void)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!cryp_pdata.reset)
+		return ret;
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1241]: prefer explicit `return TEE_SUCCESS`

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 4]
```diff
@@ -1231,6 +1233,51 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
...
+			return ret;
+
+		if (PM_HINT_IS_STATE(pm_hint, CONTEXT)) {
+			ret = stm32_cryp_reset();
+			if (ret)
+				panic();
+		}
+		break;
+	default:
+		break;
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1274]: suggestion: add `assert(0)` as this is not expected.

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 5]
```diff
@@ -1231,6 +1233,51 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
...
+		clk_disable(cryp_pdata.clock);
+		ret = TEE_SUCCESS;
+		break;
+	case PM_OP_RESUME:
+		ret = clk_enable(cryp_pdata.clock);
+		if (ret)
+			return ret;
+
+		if (PM_HINT_IS_STATE(pm_hint, CONTEXT)) {
+			ret = stm32_cryp_reset();
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1268]:
```
			if (stm32_cryp_reset())
				panic();

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 6]
```diff
@@ -1231,6 +1233,51 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
 	return res;
 }
 
+static TEE_Result stm32_cryp_reset(void)
+{
+	TEE_Result ret = TEE_SUCCESS;
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1238]: Other functions uses variable label `res` for `TEE_Result` values, prefer use `res` here also rathter than `ret`.
Ditto in `stm32_cryp_pm()` implementation.

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 7]
```diff
@@ -1229,6 +1233,51 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
...
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+
+	switch (op) {
+	case PM_OP_SUSPEND:
+		clk_disable(cryp_pdata.clock);
+		ret = TEE_SUCCESS;
+		break;
+	case PM_OP_RESUME:
+		ret = clk_enable(cryp_pdata.clock);
+		if (ret)
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1264]: I think panicking also here makes sense, does it?

File: core/drivers/crypto/stm32/stm32_cryp.c (hunk 1 above)
Comment (tbourgoi) [reply core/drivers/crypto/stm32/stm32_cryp.c:1316]: Ok, I will split the commit in 2 and push another version with 3 commits.

File: core/drivers/crypto/stm32/stm32_cryp.c (hunk 7 above)
Comment (tbourgoi) [reply core/drivers/crypto/stm32/stm32_cryp.c:1264]: Yes could be a good idea to panic, we expect to have the clock already enabled if a crypto request arrives.

**Comment by tbourgoi** [comment]
all previous comment are addressed

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 8]
```diff
@@ -1229,6 +1243,36 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
 	return res;
 }
 
+static TEE_Result stm32_cryp_pm(enum pm_op op, uint32_t pm_hint,
+				const struct pm_callback_handle *hdl __unused)
+{
+	TEE_Result res = TEE_ERROR_NOT_IMPLEMENTED;
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1249]: nitpicking suggestion:
```c
{
	switch (op) {
	case PM_OP_SUSPEND:
		clk_disable(cryp_pdata.clock);
		return TEE_SUCCESS;
	case PM_OP_RESUME:
		if (clk_enable(cryp_pdata.clock))
			panic();
		if (PM_HINT_IS_STATE(pm_hint, CONTEXT) && stm32_cryp_reset())
			panic();
		return TEE_SUCCESS;
	default:
		/* Unexpected PM operation */
		assert(0);
		return TEE_ERROR_NOT_IMPLEMENTED;
	}
}

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 9]
```diff
@@ -1286,6 +1327,9 @@ static TEE_Result stm32_cryp_probe(const void *fdt, int node,
 		}
 	}
 
+	if (IS_ENABLED(CFG_PM))
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1330]: `CFG_PM` does not exist. Can remove it.
I see that there are 2 such occurrences in stm32 drivers. We should remove them but that outside the scope of your P-R.

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 10]
```diff
@@ -1229,6 +1243,30 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
...
+		if (clk_enable(cryp_pdata.clock))
+			panic();
+
+		if (PM_HINT_IS_STATE(pm_hint, CONTEXT)) {
+			if (stm32_cryp_reset())
+				panic();
+		}
+		return TEE_SUCCESS;
+	default:
+		/* Unexpected pm operation */
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1263]: s/pm/PM/

File: core/drivers/crypto/stm32/stm32_cryp.c [hunk 11]
```diff
@@ -1229,6 +1243,30 @@ TEE_Result stm32_cryp_update(struct stm32_cryp_context *ctx, bool last_block,
...
+		clk_disable(cryp_pdata.clock);
+		return TEE_SUCCESS;
+	case PM_OP_RESUME:
+		if (clk_enable(cryp_pdata.clock))
+			panic();
+
+		if (PM_HINT_IS_STATE(pm_hint, CONTEXT)) {
+			if (stm32_cryp_reset())
+				panic();
+		}
```
Comment (etienne-lms) [inline core/drivers/crypto/stm32/stm32_cryp.c:1260]: can shrink to:
```c
		if (PM_HINT_IS_STATE(pm_hint, CONTEXT) && stm32_cryp_reset())
			panic();

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` with s/pm/PM/ addressed.

**Comment by tbourgoi** [comment]
Hello @etienne-lms 
Can I push a new branch with the fixup squashed, the last 2 modification and your tag ?

**Comment by etienne-lms** [comment]
Hello @tbourgoi, sorry I tought I had anwsered you. Yes sure, feel free to squash and apply my R-b tag.
